src/scan-imports.ts

```diff
@@ -1,159 +1,131 @@
+  const resolvedSpecifierWithoutExtension = stripJsExtension(resolvedSpecifier);
+  if (knownDependencies.includes(resolvedSpecifierWithoutExtension)) {
+    return resolvedSpecifierWithoutExtension;
   }
-  const deps = new Set<string>();
+  // Otherwise, this is an explicit import to a file within a package.
+  return resolvedSpecifier;
+}
 
-  // try to parse; this will fail if not JS
+function getInstallTargetsForFile(code: string, knownDependencies: string[]): InstallTarget[] {
+  const allImports: InstallTarget[] = [];
   try {
     const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
     traverse(ast, {
-      enter(node) {
-        const moduleName =
-          importModuleName(node as NodePath<ImportDeclaration>) ||
-          dynamicImportModuleName(node as NodePath<CallExpression>);
-
-        if (moduleName) {
-          deps.add(moduleName);
+      ImportDeclaration(path) {
+        const webModuleSpecifier = parseWebModuleSpecifier(
+          path.node.source.value,
+          knownDependencies,
+        );
+        if (webModuleSpecifier) {
+          allImports.push({
+            specifier: webModuleSpecifier,
+            all: false,
+            default: path.node.specifiers.some(s => s.type === 'ImportDefaultSpecifier'),
+            namespace: path.node.specifiers.some(s => s.type === 'ImportNamespaceSpecifier'),
+            named: path.node.specifiers
+              .map(s => s.type === 'ImportSpecifier' && s.imported.name)
+              .filter(Boolean),
+          });
+        }
+      },
+      Import(path) {
+        // Only match dynamic imports that are called as a function
+        if (path.parent.type !== 'CallExpression') {
+          return;
+        }
+        // Only match dynamic imports called with a single string argument
+        const [argNode] = path.parent.arguments;
+        if (argNode.type !== 'StringLiteral') {
+          return;
+        }
+        // Analyze that string argument as an import specifier
+        const webModuleSpecifier = parseWebModuleSpecifier(argNode.value, knownDependencies);
+        if (webModuleSpecifier) {
+          allImports.push(createInstallTarget(webModuleSpecifier, true));
         }
       },
     });
-    return [...deps];
   } catch (e) {
+    // TODO: report the error
     return [];
   }
+  return allImports;
 }
 
-interface ScanOptions {
-  dependencies: {[key: string]: string};
-  dest: string;
+export function scanWhitelist(whitelist: string[], cwd: string): InstallTarget[] {
+  const nodeModulesLoc = nodePath.join(cwd, 'node_modules');
+  return whitelist
+    .map(whitelistItem => {
+      if (!glob.hasMagic(whitelistItem)) {
+        return [createInstallTarget(whitelistItem, true)];
+      } else {
+        return scanWhitelist(glob.sync(whitelistItem, {cwd: nodeModulesLoc, nodir: true}), cwd);
+      }
+    })
+    .reduce((flat, item) => flat.concat(item));
```

Comment:
This line seems to break when I have an empty `webDependencies: []` array in `package.json`

Reply:
needed a default value for a `reduce()` call